interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,966 @@
+package requestrules
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/snapcore/snapd/interfaces/prompting"
+	"github.com/snapcore/snapd/interfaces/prompting/internal/maxidmmap"
+	"github.com/snapcore/snapd/interfaces/prompting/patterns"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+)
+
+var (
+	ErrInternalInconsistency = errors.New("internal error: prompting rules database left inconsistent")
+	ErrLifespanSingle        = errors.New(`cannot create rule with lifespan "single"`)
+	ErrRuleIDNotFound        = errors.New("rule ID is not found")
+	ErrRuleIDConflict        = errors.New("rule with matching ID already exists in rules database")
+	ErrPathPatternConflict   = errors.New("a rule with conflicting path pattern and permission already exists in the rules database")
+	ErrNoMatchingRule        = errors.New("no rules match the given path")
+	ErrUserNotAllowed        = errors.New("the given user is not allowed to request the rule with the given ID")
+)
+
+// Rule stores the contents of a request rule.
+type Rule struct {
+	ID          prompting.IDType       `json:"id"`
+	Timestamp   time.Time              `json:"timestamp"`
+	User        uint32                 `json:"user"`
+	Snap        string                 `json:"snap"`
+	Interface   string                 `json:"interface"`
+	Constraints *prompting.Constraints `json:"constraints"`
+	Outcome     prompting.OutcomeType  `json:"outcome"`
+	Lifespan    prompting.LifespanType `json:"lifespan"`
+	Expiration  time.Time              `json:"expiration,omitempty"`
+}
+
+// removePermission removes the given permission from the rule's list of
+// permissions.
+func (rule *Rule) removePermission(permission string) error {
+	if err := rule.Constraints.RemovePermission(permission); err != nil {
+		return err
+	}
+	if len(rule.Constraints.Permissions) == 0 {
+		return prompting.ErrPermissionsListEmpty
+	}
+	return nil
+}
+
+// Expired returns true if the receiving rule has a lifespan of timespan and
+// the current time is after the rule's expiration time.
+//
+// Returns an error if the rule's expiration time is invalid.
+func (rule *Rule) Expired(currentTime time.Time) (bool, error) {
+	switch rule.Lifespan {
+	case prompting.LifespanTimespan:
+		if rule.Expiration.IsZero() {
+			// Should not occur
+			return false, fmt.Errorf("encountered rule with lifespan timespan but no expiration")
```

Comment:
shouldn't all this error that should not occure be prefixed with either: "internal error:" or "unexpected error:" ?

Reply:
added that errors that looked like internal, I think I got them all.